Extract logic related with scheduler nominatedPods to an interface

Huang-Wei opened the pull request:
**What type of PR is this?**

/kind feature
/sig scheduling

**What this PR does / why we need it**:

This PR attempts to extract logic with schedulingQueue.nominatedPods as a mininum interface, so as to partly remove the dependency from preemption logic to internal schedulingQueue.

**Which issue(s) this PR fixes**:

Part of #91038.

**Special notes for your reviewer**:

This PR is for "prefactoring" purpose, so it doesn't touch the design of preemption extension point yet - i.e., doesn't pass the `podNominator` to `PostFilter()`.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

Huang-Wei commented:
/assign @alculquicondor 

It's ready for reviewing.

Huang-Wei commented:
/retest

denkensk commented:
/retest

denkensk commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -555,17 +562,16 @@ func (p *PriorityQueue) getUnschedulablePodsWithMatchingAffinityTerm(pod *v1.Pod
-	defer p.lock.RUnlock()
-	return p.nominatedPods.podsForNode(nodeName)
+	// TODO: make podsForNode() public?
+	return p.PodNominator.(*NominatedPodMap).podsForNode(nodeName)
```
Can we also add the `NominatedPodsForNode` to `PodNominator ` because it also the operation to maintain nominated Pods？

denkensk replied:
Ignore me. I find you did it in the next pr https://github.com/kubernetes/kubernetes/pull/90415

Huang-Wei replied:
Exactly, that's in #90415 :)

denkensk commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -416,14 +422,15 @@ func (p *PriorityQueue) Update(oldPod, newPod *v1.Pod) error {
 		if oldPodInfo, exists, _ := p.activeQ.Get(oldPodInfo); exists {
-			p.nominatedPods.update(oldPod, newPod)
+			// TODO: make update() public?
+			p.PodNominator.(*NominatedPodMap).update(oldPod, newPod)
```
+1 for make update() public.

ahg-g replied:
lets make it public and name it ```UpdateNominatedPod``` for now.

Huang-Wei replied:
I was hesitant on this is because the plugin developer would only need the current two methods:

```go
type PodNominator interface {
	UpdateNominatedPodForNode(pod *v1.Pod, nodeName string)
	DeleteNominatedPodIfExists(pod *v1.Pod)
}
```

If we made this update() method public, it has to be added to PodPreemptor interface, so users would be confused on which UpdateXYZ() method to use. WDYT?

Huang-Wei replied:
nvm, I renamed `UpdateNominatedPodForNode` to `AddNominatedPod`, so introducing `UpdateNominatedPod` won't be ambigurous.

denkensk commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -587,21 +593,21 @@ func (p *PriorityQueue) Close() {
+func (npm *NominatedPodMap) DeleteNominatedPodIfExists(pod *v1.Pod) {
+	npm.lock.Lock()
+	npm.delete(pod)
+	npm.lock.Unlock()
```
`defer npm.lock.Unlock()`？

Huang-Wei replied:
I usually hold using defer unless it's necessary (more than 1 return path).

ahg-g commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -711,21 +717,23 @@ func newUnschedulablePodsMap(metricRecorder metrics.MetricRecorder) *Unschedulab
 	// nominated.
 	nominatedPodToNode map[ktypes.UID]string
+
+	lock sync.RWMutex
```
It doesn't seem we need the lock. Accesses from the queue are already synced using the queue's lock. Accesses from generic_scheduler.go are not happening in parallel.

ahg-g replied:
ugh, ignore that, we need the lock to sync between accesses happening by the queue (triggered by events) and generic_scheduler.go.

I suggest we embed the lock and just use npm.Lock/Unlock.

Huang-Wei replied:
I suppose you mean embed anonymous `sync.RWMutex` and so we can use npm.Lock/Unlock?

ahg-g replied:
>I suppose you mean embed anonymous sync.RWMutex and so we can use npm.Lock/Unlock?

Yes. 

Thinking about this a bit more, I worry that we now have a second mutex, I am concerned about potential deadlocks. The original code was simpler.

How about we keep everything as is, and just use the queue itself as the backing implementation of ```PodNominator``` interface?

Huang-Wei replied:
I was thinking of that at the beginning. One primary disadvantage is that the lock is too wide - for the operation that only concern Nominating, we're locking the whole schedulingQueue, although it's the original way, but IMO is not fine-grained.

Another benefit of separate PodNominator from schedulingQueue is we can just inject a NominatedMap to the places that just need a PodNominator, like the plugins, tests, etc.

ahg-g commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -587,21 +593,21 @@ func (p *PriorityQueue) Close() {
-	p.lock.Lock()
-	p.nominatedPods.add(pod, nodeName)
-	p.lock.Unlock()
+func (npm *NominatedPodMap) UpdateNominatedPodForNode(pod *v1.Pod, nodeName string) {
```
```AddNominatedPod```

Huang-Wei replied:
done.

Huang-Wei commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -555,17 +562,16 @@ func (p *PriorityQueue) getUnschedulablePodsWithMatchingAffinityTerm(pod *v1.Pod
-	p.lock.RLock()
-	defer p.lock.RUnlock()
-	return p.nominatedPods.podsForNode(nodeName)
+	// TODO: make podsForNode() public?
```
ignore this TODO when reviewing - it's covered in the next PR.

Huang-Wei commented on pkg/scheduler/factory.go:
```diff
@@ -108,7 +108,19 @@ type Configurator struct {
 }
 
-func (c *Configurator) buildFramework(p schedulerapi.KubeSchedulerProfile) (framework.Framework, error) {
+var _ framework.PreemptHandle = &preemptHandle{}
+
+type preemptHandle struct {
+	internalqueue.NominatedPodMap
+}
+
+func newPreemptHandle() *preemptHandle {
+	return &preemptHandle{
+		*internalqueue.NewNominatedPodMap(),
+	}
+}
```
The reason is that I want to make the runtime implementation (`preemptHandle`) private. If we move to scheduling_queue.go, it has to be public. So the latter PR which does RunFilter and RunPreFilter function assignments won't work: https://github.com/kubernetes/kubernetes/pull/90711/files?file-filters%5B%5D=.go#diff-ec926856bfdbbcaf5e9dfce784909d51R165-R167.

ahg-g replied:
The other option is to put it in the framework package, and the instantiation of the preemptHandler happens inside NewFramework, and NewFramework will have an option to pass a PodNominator. This simplifies the logic significantly in this file and encapsulates it in the framework as it supposed to be.

framework.NewFramework(
  ... 
  framework.WithPodNominator(internalqueue.NewNominatedPodMap()),
)

ahg-g replied:
when we split the implementation for the framework into the ```runtime``` pkg, the preemptHandle logic will live there.

Huang-Wei replied:
SG. In this PR, I just removed the private `preemptHandle` struct, and will add it back to `framework` package the follow up PR (#90711).

ahg-g commented on pkg/scheduler/profile/profile.go:
```diff
@@ -42,8 +42,9 @@ type Profile struct {
-func NewProfile(cfg config.KubeSchedulerProfile, frameworkFact FrameworkFactory, recorderFact RecorderFactory) (*Profile, error) {
-	f, err := frameworkFact(cfg)
+func NewProfile(cfg config.KubeSchedulerProfile, frameworkFact FrameworkFactory, recorderFact RecorderFactory,
+	preemptionHelper framework.PreemptHandle) (*Profile, error) {
+	f, err := frameworkFact(cfg, preemptionHelper)
```
```suggestion
	preemptionHandle framework.PreemptHandle) (*Profile, error) {
	f, err := frameworkFact(cfg, preemptionHandle)
```

Huang-Wei replied:
done.

Huang-Wei commented:
/retest

ahg-g commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -709,21 +714,23 @@ func newUnschedulablePodsMap(metricRecorder metrics.MetricRecorder) *Unschedulab
 }
 
-// nominatedPodMap is a structure that stores pods nominated to run on nodes.
+// NominatedPodMap is a structure that stores pods nominated to run on nodes.
```
why is this exported? we just need a New function exported that returns PodNominator interface.

Huang-Wei replied:
reverted it to be private.

Huang-Wei replied:
> Errors from golint:
pkg/scheduler/internal/queue/scheduling_queue.go:807:27: exported func NewNominatedPodMap returns unexported type *queue.nominatedPodMap, which can be annoying to use

The latest commit to fix the above golint failure.

ahg-g commented on pkg/scheduler/framework/v1alpha1/interface.go:
```diff
@@ -495,3 +495,20 @@ type FrameworkHandle interface {
 }
+
+// PreemptHandle incorporates all needed logic to run preemption logic.
+// TODO: embed all preemption needed info here, and then pass it as an arg of PostFilter().
```
nit: this todo isn't really useful.

Huang-Wei replied:
removed.

Huang-Wei commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -214,14 +218,17 @@ func NewPriorityQueue(
 	}
 
+	if options.podNominator == nil {
+		options.podNominator = NewNominatedPodMap()
```
We cannot; otherwise all UTs will share the same *nominatedPodMap, which will cause UT to fail.

Huang-Wei commented:
/retest

Huang-Wei commented:
/hold
for squashing later.

ahg-g commented on pkg/scheduler/framework/v1alpha1/framework.go:
```diff
@@ -192,7 +202,9 @@ func NewFramework(r Registry, plugins *config.Plugins, args []config.PluginConfi
 		volumeBinder:          options.volumeBinder,
 		metricsRecorder:       options.metricsRecorder,
-		runAllFilters:         options.runAllFilters,
+		// TODO: inject other needed info
```
nit: this is not a useful todo

Huang-Wei replied:
removed.

ahg-g reviewed (commented):
LGTM please squash.

Huang-Wei committed:
Extract logic related with scheduler nominatedPods to an interface

- rename `UpdateNominatedPodForNode` to `AddNominatedPod`
- promote `update` to `UpdateNominatedPod`
- anonymous lock in nominatedMap
- pass PodNominator as an option to NewFramework

Huang-Wei commented:
@ahg-g squashed. PTAL.

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

ahg-g commented:
/lgtm

Huang-Wei commented:
/hold cancel
